File: src/main/java/plus/maa/backend/service/UserDetailServiceImpl.java
package plus.maa.backend.service;

import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

import lombok.RequiredArgsConstructor;
import plus.maa.backend.repository.UserRepository;
import plus.maa.backend.repository.entity.MaaUser;
import plus.maa.backend.service.model.LoginUser;

import java.util.HashSet;
import java.util.Set;

/**
 * @author AnselYuki
 */
@Service
@RequiredArgsConstructor
public class UserDetailServiceImpl implements UserDetailsService {
    private final UserRepository userRepository;

    /**
     * 查询用户信息
     *
     * @param email 用户使用邮箱登录
     * @return 用户详细信息
     * @throws UsernameNotFoundException 用户名未找到
     */
    @Override
    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
        MaaUser user = userRepository.findByEmail(email);
        if (user == null) {
            throw new UsernameNotFoundException("用户不存在");
        }

        var permissions = collectPermissionsFor(user);
        //数据封装成UserDetails返回
        return new LoginUser(user, permissions);
    }

    public Set<String> collectPermissionsFor(MaaUser user) {
        var permissions = new HashSet<String>();
        for (int i = 0; i <= user.getStatus(); i++) {
            permissions.add(Integer.toString(i));
        }
        return permissions;
    }
}

File: src/main/kotlin/plus/maa/backend/service/UserDetailServiceImpl.kt
package plus.maa.backend.service

import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service
import plus.maa.backend.repository.UserRepository
import plus.maa.backend.repository.entity.MaaUser
import plus.maa.backend.service.model.LoginUser

/**
 * @author AnselYuki
 */
@Service
class UserDetailServiceImpl(
    private val userRepository: UserRepository,
) : UserDetailsService {
    /**
     * 查询用户信息
     *
     * @param email 用户使用邮箱登录
     * @return 用户详细信息
     * @throws UsernameNotFoundException 用户名未找到
     */
    @Throws(UsernameNotFoundException::class)
    override fun loadUserByUsername(email: String): UserDetails {
        val user = userRepository.findByEmail(email) ?: throw UsernameNotFoundException("用户不存在")

        val permissions = collectAuthoritiesFor(user)
        // 数据封装成UserDetails返回
        return LoginUser(user, permissions)
    }

    fun collectAuthoritiesFor(user: MaaUser): Collection<GrantedAuthority> {
        val authorities = ArrayList<GrantedAuthority>()
        for (i in 0..user.status) {
            authorities.add(SimpleGrantedAuthority(i.toString()))
        }
        return authorities
    }
}
